## Spoznavanje z okoljem

In [47]:
import pandas as pd
import cobra

from cobra.io import load_model

import numpy as np
import pickle

from os.path import join

Preberimo model osnovnega metabolizma E. coli. Model smo prenesli s strani http://bigg.ucsd.edu/models/e_coli_core.

In [48]:
model = cobra.io.read_sbml_model(join('models','e_coli_core.xml'))

In [56]:
r_biomass = model.reactions.get_by_id('BIOMASS_Ecoli_core_w_GAM')
r_biomass

<Reaction BIOMASS_Ecoli_core_w_GAM at 0x245cd7b2e10>

In [62]:
model.objective = r_biomass

In [65]:
model.optimize()

<Solution 0.874 at 0x245dbec8710>

In [67]:
model.summary()

Rast celice je torej 0.874 $h^{-1}$

Rezultat optimizacije lahko primerjamo z orodjem [Escher-FBA](https://sbrg.github.io/escher-fba/#/app)

In [68]:
sol = model.optimize()
sol['EX_glc__D_e']

-10.0

Poskusimo spremeniti lastnosti medija. Kot vidimo zgoraj celica črpa veliko glukoze (pretok je 10 $mmol gDW^{-1} hr^{-1}$ (milimoli na gram suhe celične mase na uro - privzete enote v okolju COBRA). Kakšen je maksimalen dovod glukoze? Ali celica črpa vse kar ima na voljo?

In [75]:
r_ex_glc = model.reactions.get_by_id('EX_glc__D_e')
r_ex_glc

<Reaction EX_glc__D_e at 0x245dcd48c10>

Pretok je negativen, ampak reakcija je obrnjena! V orodju COBRA so vse *exchange* (*uptake* in *secretion*) reakcije take oblike. Kot vidimo, celica črpa vso glukozo, ki je na voljo. Poskusimo to zmanjšati.

In [76]:
r_ex_glc.bounds

(-1.0, 1000.0)

In [77]:
r_ex_glc.bounds = (0.0, 1000.0)
model.optimize()

C:\Users\miham\.conda\envs\metabolic\Lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


<Solution infeasible at 0x245dc0a5690>

Ups. Tole je bilo preveč. 

In [83]:
r_ex_glc.bounds = (-1.0, 1000.0)

In [84]:
model.optimize()

C:\Users\miham\.conda\envs\metabolic\Lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


<Solution infeasible at 0x245dc3e06d0>

Poskusimo analizo ponoviti v anaerobnih pogojih. Nastavili bomo uptake kisika.

In [80]:
r_ex_o2 = model.reactions.get_by_id('EX_o2_e')
r_ex_o2

<Reaction EX_o2_e at 0x245dcd2c350>

In [85]:
r_ex_glc.bounds = (-10.0, 1000.0)
r_ex_o2.bounds = (0, 1000)

In [86]:
model.optimize()

<Solution 0.212 at 0x245dc3e22d0>

Produkcija biomase se je zmanjšala, prav tako pa se je preusmerila pot generiranja energije. Celica je začela producira acetat (`ac`), format (`for`) in etanol (`etoh`). Prej je producirala vodo in ogljikov dioksid.

In [87]:
model.summary()

Preberimo model Recon3D. Predtem smo ga prenesli s strani https://github.com/SBRG/Recon3D. Opis modela je na voljo v [članku](http://dx.doi.org/10.1038/nbt.4072).

In [31]:
model = cobra.io.read_sbml_model('models\\Recon3D.xml')

Poglejmo nekaj atributov modela

In [7]:
model.genes[:3]

[<Gene 0 at 0x2458db6b710>,
 <Gene 26_AT1 at 0x245a0236ad0>,
 <Gene 314_AT1 at 0x245a018ac10>]

In [8]:
model.metabolites[:3]

[<Metabolite 10fthf_c at 0x245a03511d0>,
 <Metabolite 10fthf_l at 0x245a0351150>,
 <Metabolite 10fthf_m at 0x245a03512d0>]

In [5]:
model.reactions[:3]

[<Reaction 24_25DHVITD3tm at 0x1eea06c6b90>,
 <Reaction 25HVITD3t at 0x1eef3d40a50>,
 <Reaction COAtl at 0x1eea06c7210>]

Povzetek modela lahko dobimo z metodo [summary](https://cobrapy.readthedocs.io/en/latest/autoapi/cobra/summary/model_summary/index.html):

In [9]:
model.summary()

In [32]:
rx = model.reactions.get_by_id("EX_glc__D_e")

In [23]:
rx.lower_bound=0
rx.upper_bound=0

In [24]:
rx

<Reaction EX_glc__D_e at 0x245ccc12950>

In [33]:
s = model.optimize()

In [34]:
s["EX_glc__D_e"]

1000.0

In [37]:
rx

<Reaction EX_glc__D_e at 0x245dcce7a90>

In [44]:
s

<Solution 755.003 at 0x245dfb15650>

### Shadow prices and reduced costs

[Source](https://edisciplinas.usp.br/mod/resource/view.php?id=2328722)


The sensitivity of an FBA solution is indicated by either shadow prices or reduced costs.  Shadow prices are the derivative of the objective function with respect to the exchange flux of a metabolite.  They indicate how much the addition of that metabolite will increase or decrease the objective.  Reduced costs are the derivatives of the objective function with respect to an internal reaction with 0 flux, indicating how much each particular reaction affects the objective.  In the COBRA Toolbox, shadow prices and reduced costs can be calculated by optimizeCbModel.  The vector of m shadow prices is FBAsolution.y and the vector of n reduced costs is FBAsoltuion.w.


**Shadow prices**

[Figure 1](https://www.researchgate.net/publication/256452494_Flux_Imbalance_Analysis_and_the_Sensitivity_of_Cellular_Growth_to_Changes_in_Metabolite_Pools/figures?lo=1). 

Shadow prices in FBA capture the sensitivity of growth to flux imbalances. Consider the FBA problem with one metabolite and two reactions, formulated as: , ; ; . The solid red line indicates the feasible solution space, and the red dot indicates the optimal solution. When the flux balance condition is relaxed and the outgoing flux from M is allowed to increase, the feasible space moves to the right (dashed blue line) and the optimal solution increases. Since the objective function increases as the right-hand-side of the flux balance constraint decreases, the metabolite has a negative shadow price. In general for intracellular metabolites, negative shadow prices correspond to growth-limiting metabolites.

In [46]:
s.shadow_prices

10fthf_c         2.228982e-18
10fthf_l         2.228982e-18
10fthf_m         2.228982e-18
11docrtsl_c      0.000000e+00
11docrtsl_m      0.000000e+00
                     ...     
caproic_e        0.000000e+00
1a25dhvitd2_c    0.000000e+00
1a25dhvitd2_e    0.000000e+00
protein_c        0.000000e+00
h_i             -4.669125e-18
Name: shadow_prices, Length: 5835, dtype: float64

In [35]:
model.summary()

In [ ]:
rx.gene_reaction_rule

In [ ]:
for r in model.reactions[:100]:
    gpr = r.gene_reaction_rule
    if gpr:
        print(gpr)

In [ ]:
model.reactions.BIOMASS_maintenance.build_reaction_string()